jax/_src/numpy/ufuncs.py

```diff
@@ -2441,44 +2441,9 @@ def ldexp(x1: ArrayLike, x2: ArrayLike, /) -> Array:
   if (dtypes.issubdtype(x1_dtype, np.complexfloating)
       or dtypes.issubdtype(x2_dtype, np.inexact)):
     raise ValueError(f"ldexp not supported for input types {(x1_dtype, x2_dtype)}")
-
   x1, x2 = promote_shapes("ldexp", x1, x2)
-
   dtype = dtypes.canonicalize_dtype(dtypes.to_inexact_dtype(x1_dtype))
-  info = dtypes.finfo(dtype)
-  int_type = _INT_DTYPES[info.bits]
-
-  x1 = lax.convert_element_type(x1, dtype)
-  x2 = lax.convert_element_type(x2, int_type)
-
-  mask = (1 << info.nexp) - 1
-  bias = 1 - info.minexp
-  x, e = _normalize_float(x1)
-  x2 += e + ((x >> info.nmant) & mask) - bias
-
-  # find underflow/overflow before denormalization
-  underflow_cond = less(x2, -(bias + info.nmant))
-  overflow_cond = greater(x2, bias)
-
-  m = lax.full_like(x, 1, dtype=dtype)
-
-  # denormals
-  cond = less(x2, -bias + 1)
-  x2 = _where(cond, x2 + info.nmant, x2)
-  m = _where(cond, m / (1 << info.nmant), m)
-
-  x2 = lax.convert_element_type(x2, np.int32)
-  x &= ~(mask << info.nmant)
-  x |= ((lax.convert_element_type(x2, int_type) + bias) << info.nmant)
-
-  x = lax.convert_element_type(m, dtype) * lax.bitcast_convert_type(x, dtype)
-
-  # underflow
-  x = _where(underflow_cond, lax.full_like(x, 0, dtype=dtype), x)
-  # overflow
-  x = _where(overflow_cond, lax.sign(x1) * lax.full_like(x, np.inf), x)
-  # ldexp(x1, x2) = x1 for x1 = inf, -inf, nan, 0
-  return _where(isinf(x1) | isnan(x1) | (x1 == 0), x1, x)
+  return x1 * lax.convert_element_type(2 ** x2, dtype)
```

Comment:
I think this is failing because for very large `x2`, `2 ** x2` overflows when computed with with integers. We should be able to fix it by using `promote_args_inexact` instead of `promote_shapes` above, and then removing the `convert_element_type` here as it will no longer be necessary.

Reply:
@jakevdp That gives me the following error:
> jax._src.dtypes.TypePromotionError: Input dtypes ('int16', 'int32') have no available implicit dtype promotion path when jax_numpy_dtype_promotion=strict. Try explicitly casting inputs to the desired output type, or set jax_numpy_dtype_promotion=standard.

Using
```python3
x2 = lax.convert_element_type(x2, x1.dtype)
x1, x2 = promote_args_inexact("ldexp", x1, x2)
return x1 * (2 ** x2)
```
seems to fix this type error. (Replacing `promote_args_inexact` with `promote_shapes` doesn't seem to affect this.)

Reply:
Ah, yeah this is a strange case where we want to silence the normal promotion error I think. I would do something like this:
```python
x1 = promote_args_inexact("ldexp", x1)
x2 = lax.convert_element_type(x2, dtypes.dtype(x1))
return x2 * (2 ** x2)
```
Since we've asserted that `x2` is integer-typed, this will do the right thing.

Reply:
Adding a comma after `x1` and slightly editing the return to
```python3
return _where(isinf(x1) | (x1 == 0), x1, x)
```
appears to pass all tests.